drivers/input/input_ite_it8801_kbd.c

```diff
@@ -0,0 +1,219 @@
+	/* Keyboard scan out mode control register */
+	uint8_t reg_ksomcr;
+	/* Keyboard scan in data register */
+	uint8_t reg_ksidr;
+	/* Keyboard scan in edge event register */
+	uint8_t reg_ksieer;
+	/* Keyboard scan in interrupt enable register */
+	uint8_t reg_ksiier;
+};
+
+struct kbd_it8801_data {
+	struct input_kbd_matrix_common_data common;
+	struct it8801_mfd_callback it8801_kbd_callback;
+};
+
+INPUT_KBD_STRUCT_CHECK(struct kbd_it8801_config, struct kbd_it8801_data);
+
+static void kbd_it8801_drive_column(const struct device *dev, int col)
+{
+	const struct kbd_it8801_config *config = dev->config;
+	int ret;
+	uint8_t kso_val;
+
+	/* Tri-state all outputs */
+	if (col == INPUT_KBD_MATRIX_COLUMN_DRIVE_NONE) {
+		/* KSO[22:11, 6:0] output high */
+		kso_val = IT8801_REG_MASK_KSOSDIC | IT8801_REG_MASK_AKSOSC;
+	}
+	/* Assert all outputs */
+	else if (col == INPUT_KBD_MATRIX_COLUMN_DRIVE_ALL) {
+		/* KSO[22:11, 6:0] output low */
+		kso_val = IT8801_REG_MASK_AKSOSC;
+	} else {
+		/*
+		 * Selected KSO[22:11, 6:0] output low,
+		 * all others KSO output high
+		 */
+		kso_val = config->kso_mapping[col];
+	}
+
+	ret = i2c_reg_write_byte_dt(&config->i2c_dev, config->reg_ksomcr, kso_val);
+	if (ret != 0) {
+		LOG_ERR("Failed to drive column (ret %d)", ret);
+		return;
+	}
+}
+
+static kbd_row_t kbd_it8801_read_row(const struct device *dev)
+{
+	const struct kbd_it8801_config *const config = dev->config;
+	int ret;
+	uint8_t value;
+
+	ret = i2c_reg_read_byte_dt(&config->i2c_dev, config->reg_ksidr, &value);
+	if (ret != 0) {
+		LOG_ERR("Failed to read row (ret %d)", ret);
+	}
+
+	/* Bits are active-low, so invert returned levels */
+	return (~value) & 0xff;
+}
+
+void it8801_input_alert_handler(const struct device *dev)
+{
+	const struct kbd_it8801_config *const config = dev->config;
+	int ret;
+	uint8_t ksieer_val;
+
+	ret = i2c_reg_read_byte_dt(&config->i2c_dev, config->reg_ksieer, &ksieer_val);
+	if (ret != 0) {
+		LOG_ERR("Failed to read KBD interrupt status (ret %d)", ret);
+	}
+
+	if (ksieer_val != 0) {
+		/* Clear pending interrupts */
+		ret = i2c_reg_write_byte_dt(&config->i2c_dev, config->reg_ksieer, GENMASK(7, 0));
+		if (ret != 0) {
+			LOG_ERR("Failed to clear pending interrupts (ret %d)", ret);
+		}
+
+		input_kbd_matrix_poll_start(dev);
+	}
+}
+
+static void kbd_it8801_set_detect_mode(const struct device *dev, bool enable)
+{
+	const struct kbd_it8801_config *const config = dev->config;
+	int ret;
+
+	if (enable) {
+		/* Clear pending interrupts */
+		ret = i2c_reg_write_byte_dt(&config->i2c_dev, config->reg_ksieer, GENMASK(7, 0));
+		/* Enable KSI falling edge event trigger interrupt */
+		ret = i2c_reg_write_byte_dt(&config->i2c_dev, config->reg_ksiier, GENMASK(7, 0));
+	} else {
+		/* Disable KSI falling edge event trigger interrupt */
+		ret = i2c_reg_write_byte_dt(&config->i2c_dev, config->reg_ksiier, 0x00);
+	}
+
+	if (ret != 0) {
+		LOG_ERR("Failed to set detect mode (ret %d)", ret);
+		return;
+	}
```

Comment:
this should really be after each call since for the `enable == true` case you have two, just copy paste a couple times :-)

Reply:
Done.